dll/win32/shdocvw/utility.cpp

```diff
@@ -139,3 +139,89 @@ IEILIsEqual(
     FIXME("%p, %p\n", pidl1, pidl2);
     return FALSE;
 }
+
+static VOID
+SHDOCVW_PathDeleteInvalidChars(LPWSTR pszDisplayName)
+{
+#define PATH_VALID_ELEMENT ( \
+    PATH_CHAR_CLASS_DOT | PATH_CHAR_CLASS_SEMICOLON | PATH_CHAR_CLASS_COMMA | \
+    PATH_CHAR_CLASS_SPACE | PATH_CHAR_CLASS_OTHER_VALID \
+)
+    PWCHAR pch, pchSrc;
+    for (pch = pchSrc = pszDisplayName; *pchSrc; ++pchSrc)
+    {
+        if (PathIsValidCharW(*pchSrc, PATH_VALID_ELEMENT))
+            *pch++ = *pchSrc;
+    }
+    *pch = UNICODE_NULL;
+}
+
+static HRESULT
+SHDOCVW_CreateShortcut(
+    _In_ LPCWSTR pszLnkFileName, 
+    _In_ PCIDLIST_ABSOLUTE pidlTarget,
+    _In_opt_ LPCWSTR pszDescription)
+{
+    HRESULT hr;
+
+    CComPtr<IShellLink> psl;
+    hr = CoCreateInstance(CLSID_ShellLink, NULL,  CLSCTX_INPROC_SERVER,
+                          IID_PPV_ARG(IShellLink, &psl));
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+
+    psl->SetIDList(pidlTarget);
+
+    if (pszDescription)
+        psl->SetDescription(pszDescription);
+
+    CComPtr<IPersistFile> ppf;
+    hr = psl->QueryInterface(IID_PPV_ARG(IPersistFile, &ppf));
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+
+    return ppf->Save(pszLnkFileName, TRUE);
+}
+
+/*************************************************************************
+ *      AddUrlToFavorites [SHDOCVW.106]
+ */
+EXTERN_C HRESULT WINAPI
+AddUrlToFavorites(
+    _In_ HWND hwnd,
+    _In_ LPCWSTR pszUrlW,
+    _In_opt_ LPCWSTR pszTitleW,
+    _In_ BOOL fDisplayUI)
+{
+    TRACE("%p, %s, %s, %d\n", hwnd, wine_dbgstr_w(pszUrlW), wine_dbgstr_w(pszTitleW), fDisplayUI);
+
+    if (fDisplayUI)
+        FIXME("fDisplayUI\n"); // NOTE: Use SHBrowseForFolder callback
+
+    if (PathIsURLW(pszUrlW))
+        FIXME("Internet Shortcut\n");
+
+    CComHeapPtr<ITEMIDLIST> pidl;
+    HRESULT hr = SHParseDisplayName(pszUrlW, NULL, &pidl, 0, NULL);
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+
+    // Get title
+    WCHAR szTitle[MAX_PATH];
+    if (pszTitleW)
+        lstrcpynW(szTitle, pszTitleW, _countof(szTitle));
+    else
+        ILGetDisplayNameEx(NULL, pidl, szTitle, ILGDN_NORMAL);
+
+    // Delete invalid characters
+    SHDOCVW_PathDeleteInvalidChars(szTitle);
+
+    // Build shortcut pathname
+    WCHAR szPath[MAX_PATH];
+    if (!SHGetSpecialFolderPathW(hwnd, szPath, CSIDL_FAVORITES, TRUE))
+        SHGetSpecialFolderPathW(hwnd, szPath, CSIDL_COMMON_FAVORITES, TRUE);
```

Comment:
Are you sure it would go into the all-users favorites in this case? It's a bit weird that one user that happens to not have its own Favorites folder set up could be allowed to automatically save its favs in the all-users one...

Reply:
#7820